src/frontend/src/crypto/mnemonic.ts

```diff
@@ -1,18 +1,183 @@
 import { entropyToMnemonic, wordlists, validateMnemonic } from "bip39";
 import { toHexString } from "@dfinity/identity/lib/cjs/buffer";
+import { isUserNumber } from "../utils/userNumber";
 
 /**
- * @returns A random BIP39 mnemonic with 256 bits of entropy.
+ * @returns A random BIP39 mnemonic with 256 bits of entropy (generates a list of 24 words)
  */
 export function generate(): string {
-  const entropy = new Uint32Array(8);
+  const entropy = new Uint32Array(8); // NOTE: please change RECOVERYPHRASE_WORDCOUNT if this changes
   crypto.getRandomValues(entropy);
-  return entropyToMnemonic(toHexString(entropy.buffer), wordlists.english);
+  return entropyToMnemonic(
+    toHexString(entropy.buffer),
+    RECOVERYPHRASE_WORDLIST
+  );
 }
 
+/** How many words are expected in the recovery phrase */
+export const RECOVERYPHRASE_WORDCOUNT = 24;
+
+/** The list of words used to make the recovery phrase */
+// NOTE: "english" is the only one actually bundled in (see webpack config)
+export const RECOVERYPHRASE_WORDLIST: string[] = wordlists.english;
+
 /**
  * @returns true if the mnemonic is valid, false otherwise.
  */
 export function validate(mnemonic: string): boolean {
   return validateMnemonic(mnemonic);
 }
+
+/** A recovery phrase that was parsed.
+ * An object of this type should not be used to reconstruct a recovery phrase, since
+ * some information (though most likely incorrectly inserted) like extra whitespaces
+ * is lost.
+ */
+type ParsedRecoveryPhrase = {
+  leadingUserNumber: string | null;
+  words: string[];
+};
+
+/** Parse recovery phrase into a structured object */
+export const parseRecoveryPhrase = (s: string): ParsedRecoveryPhrase => {
+  /* Split nicely into words
+   * ' 2002 foo  bar ' -> split(" ") -> [ '', '2002', 'foo', '', 'bar', '' ]
+   * [ '', '2002', 'foo', '', 'bar', '' ] -> filter(...) -> [ '2002', 'foo', 'bar' ]
+   */
+  const words = s.split(" ").filter((w) => w !== "");
+
+  // check if there's a leading user (anchor) number
+  const leadingUserNumber: string | null =
+    words.length >= 1 && isUserNumber(words[0]) ? words[0] : null;
+
+  // If the first element is indeed a number (BigInt), drop it from the actual mnemonic
+  if (leadingUserNumber !== null) {
+    words.shift();
+  }
+
+  return { leadingUserNumber, words };
+};
+
+/** Drop the anchor number from a string
+ * Example: " 10005 foo bar" -> "foo bar"
+ */
+export const dropLeadingUserNumber = (input: string): string => {
+  const parsed = parseRecoveryPhrase(input);
+  return parsed.leadingUserNumber !== null
+    ? input.replace(parsed.leadingUserNumber, "").trim()
+    : input;
+};
+
+/** The warning types that may be generated. For more information on each, see 'recoveryPhraseWarnings'.
+ *
+ * We generate the messages independently mostly so that we can programatically test the warnings generated
+ * (and avoid brittle regex-based testing).*/
+export type Warning =
+  | { type: "bad_chars"; chars: string[] }
```

Comment:
I couldn't figure out how to actually add data to the enum, but maybe that's just me being stupid? I.e. don't know how to make it a proper algebraic data type

Reply:
Only the type could be an enum:

```
enum WarningType {
  BadChars,
  ExtraSpace,
  //...
}

//...
export type Warning = { type: WarningType, chars: string[] } | ...;

//...
switch type
case WarningType.BadChars
...
case WarningType.ExtraSpace
...
```

Like that it's harder to have bugs due to typos.

You can even give values to the enum:

```
enum WarningType {
  BadChars = "badChars",
....
```

Reply:
I don't think we gain that much (or at all), since (AFAIU) typescript makes all the strings (`"bad_chars"`, etc) singleton types; meaning that having a `case "bad_chors"` (notice the typo) would have typescript complain about `bad_chors` not being in the list of strings; meaning it's effectively an enum with less code -- but I'm no ts expert so I might be wrong!

Reply:
Not a strong suggestion, I was just replying to your comment.